File: zoo.py
import pandas as pd
import numpy as np

zoo=pd.read_csv("E:/Assignments/ASsignment week 12/KNN/Assignment/Zoo.csv")

def norm_func(i):
    x = (i-i.min())	/ (i.max()-i.min())
    return (x)

zoo_n = norm_func(zoo.iloc[:, 1:])
zoo_n.describe()



X = np.array(zoo_n.iloc[:,:]) # Predictors 
Y = np.array(zoo['animal name']) 

from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2)

from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors = 50)
knn.fit(X_train, Y_train)

pred = knn.predict(X_test)
pred
# Evaluate the model
from sklearn.metrics import accuracy_score
print(accuracy_score(Y_test, pred))
pd.crosstab(Y_test, pred, rownames = ['Actual'], colnames= ['Predictions'])

# error on train data
pred_train = knn.predict(X_train)
print(accuracy_score(Y_train, pred_train))
pd.crosstab(Y_train, pred_train, rownames=['Actual'], colnames = ['Predictions']) 
# creating empty list variable 
acc = []
for i in range(3,50,2):
    neigh = KNeighborsClassifier(n_neighbors=i)
    neigh.fit(X_train, Y_train)
    train_acc = np.mean(neigh.predict(X_train) == Y_train)
    test_acc = np.mean(neigh.predict(X_test) == Y_test)
    acc.append([train_acc, test_acc])
    
    
import matplotlib.pyplot as plt # library to do visualizations 

# train accuracy plot 
plt.plot(np.arange(3,50,2),[i[0] for i in acc],"ro-")

# test accuracy plot
plt.plot(np.arange(3,50,2),[i[1] for i in acc],"bo-")

File: zoo.R
library(ISLR)
library(readr)
zoo=read.csv("E:/Assignments/ASsignment week 12/KNN/Assignment/Zoo.csv")
str(zoo)
zoo$type=as.factor(zoo$type)
summary(zoo$type)
any(is.na(zoo))
zoo$animal.name=NULL
# save the Purchase column in a separate variable
purchase <- zoo[,17]

# Standarize the dataset using "scale()" R function
stzoo <- scale(zoo[,-17])
# First 100 rows for test set
test.index <- 1:20
test.data <- stzoo[test.index,]
test.purchase <- purchase[test.index]
# Rest of data for training
train.data <- stzoo[-test.index,]
train.purchase <- purchase[-test.index]
library(class)

predicted.purchase <- knn(train.data,test.data,train.purchase,k=22)
head(predicted.purchase)
mean(test.purchase != predicted.purchase)
predicted.purchase <- knn(train.data,test.data,train.purchase,k=30)
mean(test.purchase != predicted.purchase)
predicted.purchase <- knn(train.data,test.data,train.purchase,k=80)
mean(test.purchase != predicted.purchase)

predicted.purchase = NULL
error.rate = NULL

for(i in 1:20){
  
  predicted.purchase = knn(train.data,test.data,train.purchase,k=i)
  error.rate[i] = mean(test.purchase != predicted.purchase)
}
print(error.rate)
library(ggplot2)
k.values <- 1:20
error.df <- data.frame(error.rate,k.values)
error.df
ggplot(error.df,aes(x=k.values,y=error.rate)) + geom_point()+ geom_line(lty="dotted",color='red')
